### DISCUSSION TITLE
failing to find npm dependency - encoding package name

### DISCUSSION THREAD

**Discussion by mieliespoor** [renovatebot/renovate#15840] [Request Help] [answered]
Seeing the following in the logs:

```
"https://private.npm.host/npm/npm/@angular%2Fanimations (GET,404)": 1,
```
When I do the following curl requests, I get a valid response:

```
curl -s https://private.npm.host/npm/npm/%40angular/animations
curl -H 'Accept: application/json' https://private.npm.host/npm/npm/%40angular/animations
```

However, for some reason, renovate is url enconding the package name where `@angular/animations` is changed to `@angular%2Fanimations` which is resulting in the 404 being returned. How can I prevent this from happening?

**Comment by rarkins** [comment]
Encoding the `/` used to be needed for the public npmjs compatibility, but it seems like it maybe isn't now. Do either of the following work for your registry?
- `curl -s https://private.npm.host/npm/npm/%40angular%2fanimations`
- `curl -s https://private.npm.host/npm/npm/@angular/animations`

**Reply by mieliespoor** [reply]
This one  `curl -s https://private.npm.host/npm/npm/@angular/animations` works. Either with the @ represented as an @ or %40, but the / between angular and animations seems to be can't be encoded.

So this works:
```
curl -s https://private.npm.host/npm/npm/@angular/animations
curl -s https://private.npm.host/npm/npm/%40angular/animations
```

If need be, I can raise a bug with Inedo Proget to allow the encoded `/` to be accepted

**Reply by rarkins** [reply]
Are you sure it's not a reverse proxy between you and your registry changing something? Ideally it should support the same or more as `npmjs`.

**Reply by mieliespoor** [reply]
We don't have any reverse proxy's in front of the the registry, only an AWS ALB. If anything I would say it is Proget doing something funny.

**Reply by viceice** [reply]
proget is buggy. you can't use stability days with it when using it as proxy for npmjs.org. because the release timestamps are wrong.

**Reply by mieliespoor** [reply]
I did another test now. I found that it is a casing issue.

Working - `curl -s https://private.npm.host/npm/npm/@angular%2fanimations`
Not working - `curl -s https://private.npm.host/npm/npm/@angular%2Fanimations`

Notice the capital `F` in the encoding string. We are also reaching out to Proget.

**Reply by mieliespoor** [reply]
Looks to be this line: https://github.com/renovatebot/renovate/blob/main/lib/modules/datasource/npm/get.ts#L66

**Reply by rarkins** [reply]
This was modeled on `encodeURIComponent()`'s behavior, which returns `%2F` instead of `%2f`. But it's possible we could change it.

**Reply by mieliespoor** [reply]
I can create a PR to change this to to `%2f` and have you guys review. My only concern is that this might break other integrations, so I'm not sure about backward compatibility?

**Reply by rarkins** [reply]
@viceice @JamieMagee do you have any experience here?

**Reply by viceice** [reply]
not sure. can we validate how npm will do it?

**Reply by JamieMagee** [reply]
npm uses `%2f` (lowercase). Here's the relevant excerpt from `npm install @angular/animations --verbose`

```
npm http fetch GET 200 https://registry.npmjs.org/@angular%2fanimations 228ms (cache miss)
```

And the full log[^1]

[^1]: https://gist.github.com/JamieMagee/b1024b7f764be3173fb0597c9fb7d1dc

**Reply by whatatripp** [reply]
> proget is buggy. you can't use stability days with it when using it as proxy for npmjs.org. because the release timestamps are wrong.

Hi, what are "stability days" , and can you provide some detail about "release timestamps are wrong". We haven't heard about it at all before.

**Reply by viceice** [reply]
https://docs.renovatebot.com/configuration-options/#stabilitydays

https://my.inedo.com/tickets/view?ticketNumber=EDO-6213

```
===Your Name:===
Michael Kriese

===What version does this affect?===
5.2.12

===How can we help? (in brief)===
npm feed sends wrong version release dates

===Please add any details that might help us help you.===
When I access a package (eg: npm jasmine-core) over a connector, the first time i get the correct release date ("2019-09-21T00:16:24.482Z") for version 3.5.0.

When i access a second time proget uses the cached package, which sends the cache date as release date ("2019-09-24T00:38:32.307Z").

I'm using renovate bot (https://renovatebot.com/) to update my deps. It depends on release dates for correct working.
```

**Answer**
```
Hello; 
 
While we really appreciate you using our free edition, support tickets are limited to paid customers. There's also a ton of other benefits when you switch to [ProGet Enterprise](https://blog.inedo.com/proget-free-to-proget-enterprise).

But in the meantime, please post this to our public community site or Slack channel instead; our engineers and other users frequent these, and help users get the support they need. 
 
Even if this ends up being a bug in our software, the entire community will be able to benefit from our public interaction.
 
Public Q&A: https://indo.com/support/questions
 
Slack Channel: https://inedo.com/support/slack
 
NOTE: if I was incorrect, and you do actually have a paid edition of our software, please let me know. I'm sorry for the confusion!
```

**Reply by mieliespoor** [reply]
I can try and test that next week. We have a good relationship with Proget and Alana and can work with them to get this resolved. They got to this thread because of a support ticket we logged around this particular issue as well.

**Comment by mieliespoor** [comment, marked as answer]
Not sure how to create a bug from this, but I opened #15942 . I will create a PR to fix this.
